core/src/main/java/org/elasticsearch/common/lucene/ShardCoreKeyMap.java

```diff
@@ -67,15 +59,21 @@ public void add(LeafReader reader) {
         final String index = shardId.getIndex();
         synchronized (this) {
             if (coreKeyToShard.put(coreKey, shardId) == null) {
-                final boolean added = indexToCoreKey.put(index, coreKey);
+                Set<Object> objects = indexToCoreKey.get(index);
+                if (objects == null) {
+                    objects = new HashSet<>();
+                    indexToCoreKey.put(index, objects);
+                }
+                final boolean added = objects.add(coreKey);
                 assert added;
-                reader.addCoreClosedListener(new CoreClosedListener() {
-                    @Override
-                    public void onClose(Object ownerCoreCacheKey) throws IOException {
-                        assert coreKey == ownerCoreCacheKey;
-                        synchronized (ShardCoreKeyMap.this) {
-                            coreKeyToShard.remove(ownerCoreCacheKey);
-                            indexToCoreKey.remove(index, coreKey);
+                final Set<Object> coreKeys = objects;
+                reader.addCoreClosedListener(ownerCoreCacheKey -> {
+                    assert coreKey == ownerCoreCacheKey;
+                    synchronized (ShardCoreKeyMap.this) {
+                        coreKeyToShard.remove(ownerCoreCacheKey);
+                        coreKeys.remove(coreKey);
```

Comment:
I think this is OK to reuse the coreKeys reference as it is only added once, but reading this I had to double check it twice. I wonder if we should keep the old logic and use the index retrieve the correct coreKeys set, or just add a comment.